coffee/loop_scheduler.py

```diff
@@ -215,131 +215,275 @@ def merge(self, root):
 
 class ExpressionFissioner(LoopScheduler):
 
-    """Analyze data dependencies and iteration spaces, then fission associative
-    operations in expressions.
-    Fissioned expressions are placed in a separate loop nest."""
+    """Split expressions embedded in a loop nest."""
 
-    def __init__(self, cut):
+    def __init__(self, **kwargs):
         """Initialize the ExpressionFissioner.
 
-        :param cut: number of operands requested to fission expressions."""
-        self.cut = cut
+        :arg kwargs:
+            * cut: the number of operands an expression should be fissioned into
+            * match: a list of subexpressions that should be cut from the input
+                expression. ``cut`` is ignored if ``match`` is provided.
+            * loops: a value in ['all', 'expr', 'none']. 'all' means that an
+                expression is split, and its chunks are placed in separate loop
+                nests. 'expr' implies that the chunks share the outer loops of the
+                expression, particularly those not belonging to its domain. 'none'
+                means that all chunks are placed within the original loop nest
+            * perfect: if True, create perfect loop nests. This means that any
+                new loop nest in which a chunk is placed is purged from any extra
+                statement (apart, obviously, from the chunk itself)
+        """
+        self.cut = kwargs.get('cut', -1)
+        self.match = [str(i) for i in kwargs.get('match', [])]
+        self.loops = kwargs.get('loops', 'expr')
+        self.perfect = kwargs.get('perfect', False)
+
+        if 'match' in kwargs:
+            self.cutter = self.CutterMatch(self)
+        elif self.cut > 0:
+            self.cutter = self.CutterSum(self)
+        else:
+            raise RuntimeError("Must specify a `cut` or a `match`.")
+
+    class Cutter():
+
+        def __init__(self, expr_fissioner):
+            self.expr_fissioner = expr_fissioner
+
+        def cut(self, node):
+            """
+            Split ``node`` into /two halves/, called /split/ and /remainder/
+
+            For example, consider the expression a*b + c*d; if the expression is cut
+            into chunks containing only one operand (i.e., self.cut=1), then we have
+            precisely two chunks, /split/ = a*b, /remainder/ = c*d
+
+            If the input expression is a*b + c*d + e*f, and still self.cut=1, then we
+            have two chunks, /split/ = a*b, /remainder/ = c*d + e*f; that is,
+            /remainder/ always contains the subexpression after the fission point
+            """
+            self._success = False
+            left = dcopy(node)
```

Comment:
You should make `_cut` a visitor which reconstructs the tree, and then you can avoid these deep copies.